File: golang/lru.go
package main

import (
	"errors"
	"fmt"
)

type Node struct {
	key        string
	value      int
	next, prev *Node
}

func NewNode(key string, value int) *Node {
	return &Node{
		key:   key,
		value: value,
		next:  nil,
		prev:  nil,
	}
}
func NewEmptyNode() *Node {
	return &Node{
		key:   "0",
		value: 0,
		next:  nil,
		prev:  nil,
	}
}

type LinkedList struct {
	head, tail *Node
}

func NewList() *LinkedList {
	list := &LinkedList{
		head: NewEmptyNode(),
		tail: NewEmptyNode(),
	}
	list.head.next = list.tail
	list.tail.prev = list.head
	return list
}

func (self *LinkedList) insert(new_node *Node, prev *Node, curr *Node) {
	prev.next = new_node
	curr.prev = new_node
	new_node.next = curr
	new_node.prev = prev
}

func (self *LinkedList) first() *Node {
	return self.head.next
}

func (self *LinkedList) push_front(node *Node) {
	self.insert(node, self.head, self.head.next)
}

func (self *LinkedList) remove_last() *Node {
	return self.remove_node(self.tail.prev)
}

func (self *LinkedList) remove_node(ptr *Node) *Node {
	ptr.prev.next = ptr.next
	ptr.next.prev = ptr.prev
	return ptr
}

type LRUCache struct {
	heap *LinkedList
	dicc map[string]*Node
	size int
}

func NewLRUCache(size int) *LRUCache {
	return &LRUCache{
		heap: NewList(),
		dicc: make(map[string]*Node),
		size: size,
	}
}

func (self *LRUCache) Insert(key string, value int) {
	new_node := NewNode(key, value)
	if len(self.dicc) < self.size {
		self.heap.push_front(new_node)
		self.dicc[key] = new_node
	} else {
		self.heap.push_front(new_node)
		self.dicc[key] = new_node
		old_node := self.heap.remove_last()
		delete(self.dicc, old_node.key)
	}
}

func (self *LRUCache) GetValueFromKey(key string) (int, error) {
	if ptr, ok := self.dicc[key]; ok {
		self.heap.remove_node(ptr)
		self.heap.push_front(ptr)
		return ptr.value, nil
	}
	return -1, errors.New("no value for key")
}

func (self *LRUCache) GetMostRecentKey() (string, bool) {
	if len(self.dicc) > 0 {
		return self.heap.first().key, true
	}
	return "", false
}

func main() {
	fmt.Println("hello world!")

	// All operations below are performed sequentially.
	cache := NewLRUCache(3) // instantiate an LRUCache of size 3
	cache.Insert("b", 2)
	cache.Insert("a", 1)
	cache.Insert("c", 3)
	if k, ok := cache.GetMostRecentKey(); ok { //// "c" was the most recently inserted key
		fmt.Println(k)
	}
	cache.GetValueFromKey("a") //: 1
	cache.GetMostRecentKey()   // "a" // "a" was the most recently retrieved key
	cache.Insert("d", 4)       // the cache had 3 entries; the least recently
	cache.GetValueFromKey("b") // None // "b" was evicted in the previous operation
	cache.Insert("a", 5)       // "a" already exists in the cache so its valu
	cache.GetValueFromKey("a") //: 5*/
}

File: python/lru.py

class Node:
    def __init__(self, k = None, v = None):
        self.key = k
        self.value = v
        self.next = None
        self.prev = None
 

class List: 
    def __init__(self):
        self.head = Node()
        self.tail = Node()
        self.head.next = self.tail
        self.tail.prev = self.head
        self.count = 0

    def insert_between(self, prev, new_node, curr):
        prev.next = new_node
        curr.prev = new_node
        new_node.next = curr
        new_node.prev = prev
        self.count += 1

    def push_front(self, node):
        self.insert_between(self.head, node, self.head.next)

    def remove(self, node):
        node.prev.next = node.next
        node.next.prev = node.prev
        self.count -= 1
        return node

    def remove_last(self):
        return self.remove(self.tail.prev)

    def first(self):
        return self.head.next

    def last(self):

        
        return self.tail.prev

    def to_string(self):
        output = ""
        curr = self.head.next
        while curr != self.tail :
            output += "(" +  str(curr.key) + " -> " +  str(curr.value) + "), "
            curr = curr.next
        return output

#  diccionary<key, value>
class LRUCache:
    def __init__(self, size):
        self.size = size
        self.dicc = {} # <key, Node>
        self.heap = List()

    def insertKeyValuePair(self, k, v):
        if len(self.dicc) < self.size:
            current_node = Node(k, v)
            self.heap.push_front(current_node)
            self.dicc[k] = current_node
        else:
            # len(self.dicc) == self.size:
            current_node = Node(k, v)
            self.heap.push_front(current_node)

            self.dicc[k] = current_node
            old_node = self.heap.remove_last() # O(1)
            self.dicc.pop(old_node.key)

    def getMostRecentKey(self):
        if self.heap.first():
            return self.heap.first().key
        return None

    def getValueFromKey(self, k):
        if k in self.dicc:
            # TODO: update heap order
            node_to_remove = self.dicc[k] # O(1)
            self.heap.remove(node_to_remove) # O(1)
            node = Node(self.dicc[k].key, self.dicc[k].value)
            self.heap.push_front(node) # O(1)
            return node.value
        return None

# runtime O(1)
# space O(3) -> O(1)
# // All operations below are performed sequentially.
cache = LRUCache(3) # // instantiate an LRUCache of size 3
cache.insertKeyValuePair("b", 2) #: -
cache.insertKeyValuePair("a", 1) #: -
cache.insertKeyValuePair("c", 3) #: -

print(cache.getMostRecentKey()) #: "c" // "c" was the most recently inserted key

print(cache.getValueFromKey("a")) #: 1
print(cache.getMostRecentKey())# "a" // "a" was the most recently retrieved key

cache.insertKeyValuePair("d", 4) #: - // the cache had 3 entries; the least recently
print(cache.getValueFromKey("b")) #: None // "b" was evicted in the previous operation
cache.insertKeyValuePair("a", 5) # - // "a" already exists in the cache so its valu

print(cache.getValueFromKey("a")) # : 5
